File: opensource/kernel/src/main/java/com/sos/scheduler/engine/eventbus/SchedulerEventBus.java
package com.sos.scheduler.engine.eventbus;

import com.sos.scheduler.engine.data.event.Event;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class SchedulerEventBus implements EventBus, Runnable {
    private static final Logger logger = LoggerFactory.getLogger(SchedulerEventBus.class);

    private final HotEventBus hotEventBus = new HotEventBus();
    private final ColdEventBus coldEventBus = new ColdEventBus();

    @Override public void registerAnnotated(EventHandlerAnnotated o) {
        hotEventBus.registerAnnotated(o);
        coldEventBus.registerAnnotated(o);
    }

    @Override public void unregisterAnnotated(EventHandlerAnnotated o) {
        hotEventBus.unregisterAnnotated(o);
        coldEventBus.unregisterAnnotated(o);
    }

    public void register(EventSubscription s) {
        coldEventBus.register(s);
    }

    public void unregister(EventSubscription s) {
        coldEventBus.unregister(s);
    }

    public void registerHot(EventSubscription s) {
        hotEventBus.register(s);
    }

    public void unregisterHot(EventSubscription s) {
        hotEventBus.unregister(s);
    }

    public boolean isSubscribed() {
        return hotEventBus.isSubscribed() || coldEventBus.isSubscribed();
    }

    public void publish(Event e, EventSource source) {
        publish(new EventSourceEvent(e, source));
    }

    @Override public void publish(Event e) {
        logger.trace("publish {}", e);
        hotEventBus.publish(e);
        publishCold(e);
    }

    public void publishCold(Event e) {
        coldEventBus.publish(e instanceof EventSourceEvent? ((EventSourceEvent)e).getEvent() : e);
    }

    public void dispatchEvents() {
        coldEventBus.dispatchEvents();
    }

    @Override public void run() {
        coldEventBus.run();
    }

    @Override public String toString() {
        return getClass().getSimpleName();
    }
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/eventbus/SchedulerEventBus.scala
package com.sos.scheduler.engine.eventbus

import com.google.common.io.Closer
import com.sos.scheduler.engine.base.utils.ScalaUtils.RichUnitPartialFunction
import com.sos.scheduler.engine.common.scalautil.Closers.implicits._
import com.sos.scheduler.engine.common.scalautil.Logger
import com.sos.scheduler.engine.data.event.{AnyKeyedEvent, Event, KeyedEvent, NoKeyEvent}
import com.sos.scheduler.engine.eventbus.SchedulerEventBus._
import scala.reflect.ClassTag

final class SchedulerEventBus extends EventBus with Runnable {

  private val hotEventBus = new HotEventBus
  private val coldEventBus = new ColdEventBus

  def registerAnnotated(o: EventHandlerAnnotated): Unit = {
    hotEventBus.registerAnnotated(o)
    coldEventBus.registerAnnotated(o)
  }

  def unregisterAnnotated(o: EventHandlerAnnotated): Unit = {
    hotEventBus.unregisterAnnotated(o)
    coldEventBus.unregisterAnnotated(o)
  }

  /**
   * Calls methods `onHotBasic` and `on`.
   * `eventHandler` will be called twice: synchronously at the event and later asynchronously.
   * @param eventHandler Events not in the function's domain are ignored.
   */
  def onHotAndCold[E <: Event: ClassTag](eventHandler: PartialFunction[KeyedEvent[E], Unit])(implicit closer: Closer): Unit = {
    onHot[E](eventHandler)
    on[E](eventHandler)
  }

  /**
   * Subscribes events of class E for eventHandler and registers this in [[Closer]] for unsubscription with Closer.closer().
   * The event handler is called asynchronous by dispatchEvent().
   * @param eventHandler Events not in the function's domain are ignored.
   */
  def on[E <: Event](eventHandler: PartialFunction[KeyedEvent[E], Unit])(implicit closer: Closer, e: ClassTag[E]): Unit =
    subscribeKeyedClosable[E](coldEventBus, eventHandler)

  /**
   * Subscribes events of class E for eventHandler and registers this in [[Closer]] for unsubscription with Closer.closer().
   * The event handler is synchronous.
   * @param eventHandler Events not in the function's domain are ignored.
   */
  def onHot[E <: Event: ClassTag](eventHandler: PartialFunction[KeyedEvent[E], Unit])(implicit closer: Closer): Unit =
    subscribeKeyedClosable[E](hotEventBus, eventHandler)

  private def subscribeKeyedClosable[E <: Event: ClassTag](whichEventBus: EventBus, handleEvent: PartialFunction[KeyedEvent[E], Unit])
    (implicit closer: Closer): Unit
  =
    subscribeClosable[E](whichEventBus, EventSubscription[E] {
      case e ⇒ handleEvent.callIfDefined(e)
    })

  private def subscribeClosable[E <: Event: ClassTag](whichEventBus: EventBus, subscription: EventSubscription)(implicit closer: Closer): Unit = {
    whichEventBus.subscribe(subscription)
    closer.onClose {
      whichEventBus.unsubscribe(subscription)
    }
  }
  def subscribe(s: EventSubscription): Unit = coldEventBus.subscribe(s)

  def unsubscribe(s: EventSubscription): Unit = coldEventBus.unsubscribe(s)

  def subscribeHot(s: EventSubscription): Unit = hotEventBus.subscribe(s)

  def unsubscribeHot(s: EventSubscription): Unit = hotEventBus.unsubscribe(s)

  def isSubscribed = hotEventBus.isSubscribed || coldEventBus.isSubscribed

  def publish(e: NoKeyEvent): Unit = publish(KeyedEvent(e))

  def publish(e: AnyKeyedEvent): Unit = {
    logger.trace(s"Publish $e")
    hotEventBus.publish(e)
    publishCold(e)
  }

  def publishCold(keyedEvent: AnyKeyedEvent): Unit =
    coldEventBus.publish(keyedEvent)

  def dispatchEvents(): Unit = coldEventBus.dispatchEvents()

  def run(): Unit = coldEventBus.run()

  override def toString = getClass.getSimpleName
}

object SchedulerEventBus {
  private final val logger = Logger(getClass)
}
